aiutils/data/data.py

```diff
@@ -0,0 +1,87 @@
+"""Utility functions for making batch generation easier."""
+import numpy as np
+from multiprocessing import Process, Queue
+
+
+def sequential(batch_size, num_samples, num_epochs=1):
+    """Generate sequence indices.
+    """
+    for epoch in range(num_epochs):
+        indices = np.arange(num_samples)
+        for i in range(0, num_samples - batch_size + 1, batch_size):
+            slice_ = slice(i, i + batch_size)
+            yield indices[slice_]
+
+
+def random(batch_size, num_samples, num_epochs=1):
+    """Generate random indices.
+    """
+    for epoch in range(num_epochs):
+        indices = np.random.permutation(num_samples)
+        for i in range(0, num_samples - batch_size + 1, batch_size):
+            slice_ = slice(i, i + batch_size)
+            yield indices[slice_]
+
+
+def batch_generator(data, index_generator, batch_function=None):
+    """Generate batches of data.
+    """
+    for samples in index_generator:
+        batch = data.get(samples)
+        if batch_function:
+            output = batch_function(batch)
+        else:
+            output = batch
+        yield output
+
+
+def async_batch_generator(data,
+                          index_generator,
+                          queue_maxsize,
+                          batch_function=None):
+    """Create an asynchronous batch generator.
+    """
+    batcher = batch_generator(data, index_generator, batch_function)
+
+    queue = Queue(maxsize=queue_maxsize)
+    fetcher = BatchFetcher(queue, batcher)
+    fetcher.start()
+
+    queue_batcher = queue_generator(queue)
+    return queue_batcher
+
+
+def queue_generator(queue, sentinel=None):
+    """Create a generator from a queue.
+    """
+    while True:
+        value = queue.get()
+        if value is not sentinel:
+            yield value
+        else:
+            return
+
+
+class BatchFetcher(Process):
+    def __init__(self, queue, batch_generator):
+        super(BatchFetcher, self).__init__()
```

Comment:
i know python doc recommends super instead of Process.__init__() but i am not very sure about why?